Add error message for non-existent staging project

saraycp opened the pull request:
Before, it was raising an exception when the staging project didn't exist. 

```
osc api -X GET '/staging/openSUSE:Factory/staging_projects/openSUSE:Factory:Staging:Fake'                                             
      
=>                                                                                
Server returned an error: HTTP Error 404: Not Found                                  
Couldn't find Project with [WHERE (projects.id not in (0)) AND `projects`.`staging_wor
kflow_id` = ? AND `projects`.`name` = ?]
```

Now, it shows a proper error.

```
osc api -X GET '/staging/openSUSE:Factory/staging_projects/openSUSE:Factory:Staging:Fake' 

=>
Server returned an error: HTTP Error 404: Not Found
Staging Project openSUSE:Factory:Staging:Fake not found.
```

Fixes: #8700 and also fixes: #8686

<!---
If you haven't done so already, please read the CONTRIBUTING.md file to learn
how we work and what we expect from all contributors.

https://github.com/openSUSE/open-build-service/blob/master/CONTRIBUTING.md

In order to make it as easy as possible for other developers to review your
pull request we ask you to:

- Explain what this PR is about in the description
- Explain the steps the reviewer has to follow to verify your change
- If the reviewer needs sample data to verify your change, please explain how to
  create that data
- If you include visual changes in this PR, please add screenshots or GIFs
- If you address performance in this PR, add benchmark data or explain how the
  reviewer can benchmark this

This is a good PR description example:

Hey Friends,

this introduces labels for the different build result states on the project
monitor page. This makes it easier to get a visual overview of what is going on
in your project.

To verify this feature

- Enable the interconnect to build.opensuse.org
- Create the project home:Admin
- Add 'openSUSE Tumbleweed' as a repository to the project
- Branch a couple of packages into the project:
  ```
  for i in `osc -A http://0.0.0.0:3000 ls openSUSE.org:home:hennevogel`; do osc -A http://0.0.0.0:3000 copypac openSUSE.org:home:hennevogel $i home:Admin; done
  ```
- Visit the monitor page and see the new labels for the different states.

Here is a screenshot of how it looks:

** Before **
![Screenshot of the project monitor](https://example.com/screenshot1.png)

** After **
![Screenshot of the project monitor](https://example.com/screenshot2.png)

-->

dmarcoux commented:
It's all good, I pushed a commit to test the changes instead of commenting about adding a spec. 

The only thing left is to squash the commits.

coolo commented:
But why does build.opensuse.org throw a 500 then?

coolo commented:
Oh, possibly because I'm admin on the one instance and normal user on the other?

saraycp commented:
> It's all good slightly_smiling_face
> 
> I pushed a commit to test the changes instead of commenting about adding a spec.
> 
> The only thing left is to squash the commits.

Thanks @dmarcoux, squashed!

coolo commented:
Ah, no. So sorry @saraycp but this is not fixing the issue - I just mixed up two things in that issue.

```
coolo@f102#api>tosc api -X GET '/staging/openSUSE:Factory/staging_projects/openSUSE:Factory:Staging:Fake'  
Server returned an error: HTTP Error 404: Not Found
Couldn't find Project with [WHERE (projects.id not in (0)) AND `projects`.`staging_workflow_id` = ? AND `projects`.`name` = ?]
coolo@f102#api>tosc api -X GET '/staging/home:coolo/staging_projects/openSUSE:Factory:Staging:Fake'  
Server returned an error: HTTP Error 500: Internal Server Error
```

You fix the rather strange error message in here, but the issue is about the crash on checking the route on a project without staging workflow. Just didn't realize that I mixed 15.1 and 15.2 when trying obs :)

saraycp commented:
> Ah, no. So sorry @saraycp but this is not fixing the issue - I just mixed up two things in that issue.

No worries, I'll create another issue just for this.

saraycp commented:
Fixes: #8700

vpereira commented on src/api/app/controllers/staging/staging_projects_controller.rb:
```diff
@@ -20,7 +20,14 @@ def index
   end
 
   def show
-    @staging_project = @project.staging.staging_projects.find_by!(name: params[:staging_project_name])
```
Hm, isn't that better to raise `ActiveRecord::RecordNotFound` and catch it on `application_controller` ? That's, IMO, the "OBS way" to handle it.

coolo replied:
It's already caught there. That's why you get a 404 and not a 500. It's just that `RecordNotFound` renders this rather strange error. So to fix it 'the OBS way', you would have to raise e.g. InvalidProjectParameters

vpereira replied:
@coolo I'm fine with that, even before if we go DRY and use the same code doing that for `show` and `accept`, or do you think that's better to do the error handling in every single action?

coolo replied:
I don't care either way. I'm just saying that your review comment makes little sense as the `find_by!` that @saraycp patches out is doing exactly what you're suggesting - including the creation of the issue.

So custom error message is all that's required - either through render_error or raise. I prefer raise, but exceptions aren't everyone's darling.

saraycp replied:
@vpereira I understand your point, the DRYer the better. But this is an exceptional case and  we just want to customize the message at this point for now.

I've force-pushed already to use `raise` instead of `render_error`, halfway between both ideas.

vpereira replied:
> don't care either way. I'm just saying that your review comment makes little sense as the `find_by!` that @saraycp patches out is doing exactly what you're suggesting

it was a rhetoric question @coolo ;)

@saraycp I don't see why we should move away from the `bang` finder methods.. any reason?  I had the impression, based in past reviews, that it was the rule here.. BTW I think the action `accept` has the same problem right? So maybe a `before_action` taking care of that for both methods - raising something like https://github.com/openSUSE/open-build-service/blob/00d38f203dd9469c132a3abb3de00fb2525946ad/src/api/app/controllers/source/errors.rb would be, IMO,  a better approach

coolo replied:
I personally find the proposed solution pretty brilliant - no need for a different exception. And the bang method throws that error the issue is about - you can't pass that message to the exception as @saraycp does here, so I see no way not to move away from them *and* having custom errors.

coolo commented:
8704 is for the webui and requires different mechanics.

saraycp commented:
> 8704 is for the webui and requires different mechanics.

Exactly! That was Webui, I just had to redirect a set a flash message when something wrong happend.

This PR is related to API and I planned to provide a solution taking into account your suggestions, @vpereira. For this we are not in a hurry, so I add WIP.

saraycp commented:
@vpereira, I've added a commit trying to follow your proposal. Am I in the right direction?

A few things to have into account:

- The `find_by!` we had in `accept` weren't exactly the same as the one in `show` action. I've unified and moved them to a `before_action`. I hope it doesn't make any difference.
- Calling `find_by!`, as usual, doesn't allow us to customize the error message. So I have to call `find_by` instead and raise an exception with a customized message.
- I created a new exception in error.rb and raise it when the staging project doesn't exist. *

* I created and exception but I still don't see the benefit of doing this over raising an already existent exception like 'ActiveRecord::RecordNotFound'.

coolo commented on src/api/app/controllers/staging/staging_projects_controller.rb:
```diff
@@ -69,4 +69,10 @@ def set_options
   end
+
+  def set_staging_project
+    @staging_project = @project.staging.staging_projects.find_by(name: params[:staging_project_name])
```
now also raise  *NotFound `unless @project.staging` to fix #8686 - you're so close :)

saraycp replied:
Just to clarify: it raised a 500 error when the project did exist but it wasn't associated to a staging workflow. If the project didn't exist, it was raising a proper 404 error.

Now the first case is also raising a 404. Please have another look, @coolo.

coolo commented on src/api/app/controllers/source/errors.rb:
```diff
@@ -94,4 +94,8 @@ class ProjectCopyNoPermission < APIError
   class InvalidFlag < APIError; end
+
+  class StagingProjectNotFound < APIError
+    setup 'not_found', 404
```
btw: if you just setup 404, the error will be `staging_project_not_found` - and this is basically the only benefit of not using the generic `ActiveRecord` exception. The class name of the exception becomes the user facing error.

coolo replied:
Really, leaving the 'not_found' improves the API usage - a little bit. API errors follow e.g. this format:

```
osc -d api -X PUT -T Rakefile /source/GNOME:Factory/gnome-shell/Rakefile
...
<status code="put_file_no_permission">
   <summary>Insufficient permissions to store file in package gnome-shell, project GNOME:Factory</summary></status>
```

This maps to `raise PutFileNoPermission, "Insufficient permissions to store file in package #{@package_name}, project #{@project_name}"`

So with 'not_found', it says code='not_found', without it it says code='staging_workflow_not_found' or code='staging_project_not_found' - and the client has an easy way to differ. And you have to type less.

But I don't think it's important enough to hold it back more.

saraycp commented:
Fixes: #8700 and also fixes: #8686
